Remotion/Validation/UnitTests/Validators/NotEmptyOrWhitespaceValidatorTest.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Linq;
+using NUnit.Framework;
+using Remotion.Validation.Implementation;
+using Remotion.Validation.Results;
+using Remotion.Validation.Validators;
+
+namespace Remotion.Validation.UnitTests.Validators
+{
+  [TestFixture]
+  public class NotEmptyOrWhitespaceValidatorTest : ValidatorTestBase
+  {
+    [Test]
+    public void Validate_WithPropertyValueNull_ReturnsNoValidationFailures ()
+    {
+      var propertyValidatorContext = CreatePropertyValidatorContext(null);
+      var validator = new NotEmptyOrWhitespaceValidator(new InvariantValidationMessage("Fake Message"));
+
+      var validationFailures = validator.Validate(propertyValidatorContext);
+
+      Assert.That(validationFailures, Is.Empty);
+    }
+
+    [Test]
+    public void ValidateString_WithPropertyValueIsEmptyString_ReturnsSingleValidationFailure ()
+    {
+      var propertyValidatorContext = CreatePropertyValidatorContext("");
+      var validator = new NotEmptyOrWhitespaceValidator(new InvariantValidationMessage("Custom validation message."));
+
+      var validationFailures = validator.Validate(propertyValidatorContext).ToArray();
+
+      AssertValidationFailure(validationFailures, propertyValidatorContext);
+    }
+
+    [Test]
+    public void ValidateString_WithPropertyValueIsLetter_ReturnsNoValidationFailures ()
+    {
+      var propertyValidatorContext = CreatePropertyValidatorContext("y");
+      var validator = new NotEmptyOrWhitespaceValidator(new InvariantValidationMessage("Fake Message"));
+
+      var validationFailures = validator.Validate(propertyValidatorContext);
+
+      Assert.That(validationFailures, Is.Empty);
+    }
+
+    [Test]
+    public void ValidateString_WithPropertyValueHasLeadingAndTrailingWhitespace_ReturnsNoValidationFailures ()
+    {
+      var propertyValidatorContext = CreatePropertyValidatorContext("\ty  ");
+      var validator = new NotEmptyOrWhitespaceValidator(new InvariantValidationMessage("Fake Message"));
+
+      var validationFailures = validator.Validate(propertyValidatorContext);
+
+      Assert.That(validationFailures, Is.Empty);
+    }
+
+    [Test]
+    public void ValidateString_WithPropertyValueIsWhitespaceString_ReturnsSingleValidationFailure ()
+    {
+      var whitespaceString = " \t\r\n \u00a0";
```

Comment:
Zero-width space is not a whitespace character, but a format character: https://github.com/dotnet/runtime/issues/29613
The list of charaters in the "separator, space" category contains "Thin space" (U+2009) and "Hair Space" (U+200B), but not "Zero width space" (U+200A), which is in the category "other, format"

https://unicode.org/reports/tr44/ says that "For example, U+200B ZERO WIDTH SPACE was originally classified as a space character (General_Category=Zs), but it was reclassified as a Format character (General_Category=Cf) to clearly distinguish it from space characters in its function as a format control for line breaking."

How should we deal with this?

Reply:
do you want to add a dedicated test to document such non-obvious cases or just leave it to the docs/wiki/specs since we don't actually have code for it?

pro test: this is going to come up in the future.

pro specs only: it's not this code's concern.

personal experience: non-obvious things do come up as a question later on and a dedicated testcase for the most-likely special cases does help in finding an answer fast.

Reply:
Test for zero-width space added.